File: csharp/0063-unique-paths-ii.cs
public class Solution {
    public int UniquePathsWithObstacles(int[][] obstacleGrid) {
        var rowLength = obstacleGrid.Length;
        var colLength = obstacleGrid[0].Length;

        int[] row = new int[colLength];
        Array.Fill(row, 0);
        row[colLength - 1] = 1;

        for(int i = rowLength - 1; i >= 0; i--)
        {
            for(int j = colLength - 1; j >= 0; j--)
            {
                if(obstacleGrid[i][j] == 1)
                    row[j] = 0;
                else if (j + 1 < colLength) 
                    row[j] = row[j] + row[j + 1];
            }
        }
        return row[0];
    }
}

File: rust/0063-unique-paths-ii.rs
impl Solution {
    // Time O(m*n) - Space O(n)
    pub fn unique_paths_with_obstacles(obstacle_grid: Vec<Vec<i32>>) -> i32 {
        let n = obstacle_grid[0].len();
        let mut dp = vec![0; n + 1];
        dp[1] = 1;
        for row in obstacle_grid {
            for i in 1..dp.len() {
                if row[i - 1] == 1 {
                    dp[i] = 0;
                } else {
                    dp[i] += dp[i - 1];
                }
            }
        }
        *dp.last().unwrap()
    }
}
